libraries/AP_Proximity/AP_Proximity_Boundary_3D.cpp

```diff
@@ -0,0 +1,287 @@
+#include "AP_Proximity_Backend.h"
+#include "AP_Proximity_Boundary_3D.h"
+
+/*
+  Constructor. 
+  This incorporates initialisation as well.
+*/
+AP_Proximity_Boundary_3D::AP_Proximity_Boundary_3D() 
+{
+    // initialise sector edge vector used for building the boundary fence
+    init();
+}
+
+// initialise the boundary and sector_edge_vector array used for object avoidance
+//   should be called if the sector_middle_deg or _sector_width_deg arrays are changed
+void AP_Proximity_Boundary_3D::init()
+{
+    for (uint8_t layer=0; layer < PROXIMITY_NUM_LAYERS; layer++) {
+        const float pitch = ((float)_pitch_middle_deg[layer]);
+        for (uint8_t sector=0; sector < PROXIMITY_NUM_SECTORS; sector++) {
+            const float angle_rad = ((float)_sector_middle_deg[sector]+(PROXIMITY_SECTOR_WIDTH_DEG/2.0f));
+            _sector_edge_vector[layer][sector].offset_bearing(angle_rad, pitch, 100.0f);
+            _boundary_points[layer][sector] = _sector_edge_vector[layer][sector] * PROXIMITY_BOUNDARY_DIST_DEFAULT;
+        }
+    }
+}
+
+// returns face corresponding to the provided yaw and (optionally) pitch
+// pitch is the vertical body-frame angle (in degrees) to the obstacle (0=directly ahead, 90 is above the vehicle)
+// yaw is the horizontal body-frame angle (in degrees) to the obstacle (0=directly ahead of the vehicle, 90 is to the right of the vehicle)
+AP_Proximity_Boundary_3D::Face AP_Proximity_Boundary_3D::get_face(float pitch, float yaw) const
+{   
+    const uint8_t sector = wrap_360(yaw + (PROXIMITY_SECTOR_WIDTH_DEG * 0.5f)) / 45.0f;
+    const float pitch_limited = constrain_float(pitch, -75.0f, 74.9f);
```

Comment:
I made it 74.9 because 75 translates to the begining "6 th" layer.. whereas we can only handle 5 layers atm. 0 being the bottom most, 2 being the middle, and 5 being the top most.